K2Field.K2NE.ServiceBroker/ServiceObjects/ExcelServices/ExcelImportServiceSO.cs

```diff
@@ -0,0 +1,528 @@
+                                    }
+                                    else if (nonblankCharCount > 0 && newSmartObject.Properties[sColName].Type == SOC.PropertyType.Time) // Time column
+                                    {
+                                        tmpDate = DateTime.FromOADate(Convert.ToDouble(dr[sColName].ToString()));
+                                        newSmartObject.Properties[sColName].Value = tmpDate.ToShortTimeString();
+                                    }
+                                    else // not a Date or DateTime column
+                                    {
+                                        newSmartObject.Properties[sColName].Value = dr[sColName].ToString();
+                                    }
+                                }
+
+                                // Add the transaction ID value for identification purposes.
+                                if (sTransactionIDName != string.Empty && sTransactionIDValue != string.Empty)
+                                {
+                                    // replace spaces with underscores as SmartObject system names do that
+                                    newSmartObject.Properties[sTransactionIDName.Replace(" ", "_")].Value = sTransactionIDValue;
+                                }
+
+                                if (!string.IsNullOrEmpty(FQN))
+                                {
+                                    newSmartObject.Properties["FQN"].Value = FQN;
+                                }
+
+                                inputList.SmartObjectsList.Add(newSmartObject);
+                            }
+                            try
+                            {
+                                smoClientServer.ExecuteBulkScalar(soImport, inputList);
+                            }
+                            catch (SOC.SmartObjectException ex)
+                            {
+                                result = ex.Message + ex.StackTrace;
+                                foreach (SOC.SmartObjectExceptionData smOExBrokerData in ex.BrokerData)
+                                {
+                                    result += smOExBrokerData.Message + "  Before failure, " + result;
+                                }
+                                return result;
+                            }
+
+                            // Indicate the Transaction ID name and value in the results field
+                            if (sTransactionIDName != string.Empty && sTransactionIDValue != string.Empty)
+                            {
+                                result += ". " + sTransactionIDName + ":" + sTransactionIDValue;
+                            }
+                        }
+                        // free objects
+                        arrMatchingCols = null;
+                        soImport = null;
+                    }
+                    catch (ApplicationException ex)
```

Comment:
The overall operation of this service broker is that if something fails that we throw the exception all the way up to the highest level and let it be handled there in the Execute() method.

It looks like here, we're trying to be more graceful but i would suggest to not do that and just raise the exception to the exceute and let people deal with it on the layer that calls the SMO.

Reply:
Done.